src/send/SendSelectRecipient.tsx

```diff
@@ -135,27 +152,167 @@ const getStartedStyles = StyleSheet.create({
   },
 })
 
-function SendSelectRecipient() {
+function SendSelectRecipient({ route }: Props) {
   const { t } = useTranslation()
 
+  const forceTokenId = route.params?.forceTokenId
+  const defaultTokenIdOverride = route.params?.defaultTokenIdOverride
+
   const [searchQuery, setSearchQuery] = useState('')
-  const throttledSearch = throttle((searchInput: string) => {
-    setSearchQuery(searchInput)
-  }, SEARCH_THROTTLE_TIME)
+  const [showSearchResults, setShowSearchResults] = useState(false)
+
+  const defaultCountryCode = useSelector(defaultCountryCodeSelector)
+  const [sendOrInviteButtonHidden, setSendOrInviteButtonHidden] = useState(true)
+
+  // We debounce the search query in order to perform network calls to check
+  // if the query resolves to some special sort of identifier.
+  const [debouncedSearchQuery, setDebouncedSearchQuery] = useState(searchQuery)
+  const debounceSearchQuery = useCallback(
+    debounce((query: string) => {
+      setDebouncedSearchQuery(query)
+    }, TYPING_DEBOUNCE_MILLSECONDS),
+    []
+  )
+  useEffect(() => {
+    const parsedPhoneNumber = parsePhoneNumber(
+      searchQuery,
+      defaultCountryCode ? defaultCountryCode : undefined
+    )
+
+    if (parsedPhoneNumber) {
+      debounceSearchQuery(parsedPhoneNumber.e164Number)
+    } else {
+      debounceSearchQuery(searchQuery)
+    }
+  }, [searchQuery, defaultCountryCode])
+  const { result: resolveAddressResult } = useAsync(resolveId, [debouncedSearchQuery])
+
+  const setSearchQueryWrapper = (query: string) => {
+    // Always unset the selected recipient and hide the send/invite button
+    // when the search query is changed in order to prevent edge cases
+    // where the button appears but is bound to a recipient that is
+    // not present on the page.
+    unsetSelectedRecipient()
+    setSendOrInviteButtonHidden(true)
+
+    if (!query) {
+      setShowSearchResults(false)
+    } else {
+      setShowSearchResults(true)
+    }
+    setSearchQuery(query)
+  }
+
+  const [showInviteModal, setShowInviteModal] = useState(false)
 
   const [showContacts, setShowContacts] = useState(false)
 
   const recentRecipients = useSelector((state) => state.send.recentRecipients)
   const contactsCache = useSelector(phoneRecipientCacheSelector)
+  const recipientInfo = useSelector(recipientInfoSelector)
+
   const contactRecipients = useMemo(
     () => sortRecipients(Object.values(contactsCache)),
     [contactsCache]
   )
+  const [contactsFiltered, setContactsFiltered] = useState(() =>
+    sortRecipients(Object.values(contactRecipients))
+  )
+  const [recentFiltered, setRecentFiltered] = useState(() => recentRecipients)
+
+  const recentRecipientsFilter = useMemo(
```

Comment:
these are defined pretty far away from where they are used, which makes it harder to understand what they are supposed to do, and harder to understand the later code that relies on them

Reply:
by these I mean this and contactRecipientsFilter, though there's probably more examples